qa/tasks/cephfs/test_quiesce.py

```diff
@@ -0,0 +1,627 @@
+
+        P = self.fs.rank_tell(["ops"])
+        log.debug(f"{P}")
+
+    def test_quiesce_path_link_terminal(self):
+        """
+        That quiesce on path with an terminal link fails with ENOTDIR even
+        pointing to a valid subvolume.
+        """
+
+        self._configure_subvolume()
+
+        self.mount_a.run_shell_payload("ln -s ../.. subvol_quiesce")
+        path = self.mount_a.cephfs_mntpt + "/subvol_quiesce"
+
+        J = self.fs.rank_tell(["quiesce", "path", path, '--wait'])
+        log.debug(f"{J}")
+        self.assertEqual(J['op']['result'], -20) # ENOTDIR: the link is not a directory
+
+    def test_quiesce_path_link_intermediate(self):
+        """
+        That quiesce on path with an intermediate link fails with ENOTDIR.
+        """
+
+        self._configure_subvolume()
+
+        self.mount_a.run_shell_payload("ln -s ../../.. _nogroup")
+        path = self.mount_a.cephfs_mntpt + "/_nogroup/" + self.QUIESCE_SUBVOLUME
+
+        J = self.fs.rank_tell(["quiesce", "path", path, '--wait'])
+        log.debug(f"{J}")
+        self.assertEqual(J['op']['result'], -20) # ENOTDIR: path_traverse: the intermediate link is not a directory
+
+    def test_quiesce_path_notsubvol(self):
+        """
+        That quiesce on a directory under a subvolume is valid.
+        """
+
+        self._configure_subvolume()
+
+        self.mount_a.run_shell_payload("mkdir dir")
+        path = self.mount_a.cephfs_mntpt + "/dir"
+
+        J = self.fs.rank_tell(["quiesce", "path", path, '--wait'])
+        reqid = self.reqid_tostr(J['op']['reqid'])
+        self._wait_for_quiesce_complete(reqid)
+        self._verify_quiesce(root=path)
+
+    def test_quiesce_path_regfile(self):
+        """
+        That quiesce on a regular file fails with ENOTDIR.
+        """
+
+        self._configure_subvolume()
+
+        self.mount_a.run_shell_payload("touch file")
+        path = self.mount_a.cephfs_mntpt + "/file"
+
+        J = self.fs.rank_tell(["quiesce", "path", path, '--wait'])
+        log.debug(f"{J}")
+        self.assertEqual(J['op']['result'], -20) # ENOTDIR
+
+    def test_quiesce_path_dup(self):
+        """
+        That two identical quiesce ops will result in one failing with
+        EINPROGRESS.
+        """
+
+        self._configure_subvolume()
+
+        op1 = self.fs.rank_tell(["quiesce", "path", self.subvolume])['op']
+        op1_reqid = self.reqid_tostr(op1['reqid'])
+        op2 = self.fs.rank_tell(["quiesce", "path", self.subvolume, '--wait'])['op']
+        op1 = self.fs.get_op(op1_reqid)['type_data'] # for possible dup result
+        log.debug(f"op1 = {op1}")
+        log.debug(f"op2 = {op2}")
+        self.assertIn(op1['flag_point'], (self.FP_QUIESCE_COMPLETE, 'cleaned up request'))
+        self.assertIn(op2['flag_point'], (self.FP_QUIESCE_COMPLETE, 'cleaned up request'))
+        self.assertTrue(op1['result'] == -115 or op2['result'] == -115) # EINPROGRESS
```

Comment:
```suggestion
        self.assertTrue(op1['result'] == -errno.EINPROGRESS or op2['result'] == -errno.EINPROGRESS, 
            'op1: {[result]}, op2: {[result]}, EINPROGRESS: {}'.format(op1, op2, errno.EINPROGRESS)
        )
```

Reply:
ceph returns OS-local definitions of the errors, so we should use the defines. On MacOS `EINPROGRESS` is 36

Reply:
MDS should return an errno standardized in `src/include/fs_types.h`. In practice, these values all come from linux (to maintain compatibility with existing clients/applications).

So the test will remain the same but the C++ code should work correctly for you now.

Reply:
See also: #55647

Reply:
that's not how I understood it should work. The admin socket response usually flows through the MCommandReply which performs host->ceph(linux) during encoding/decoding

```
  void encode(ceph::buffer::list &bl) const {
    using ceph::encode;
    __s32 newcode = hostos_to_ceph_errno(code);
    encode(newcode, bl);
  }
  void decode(ceph::buffer::list::const_iterator &bl) {
    using ceph::decode;
    decode(code, bl);
    code = ceph_to_hostos_errno(code);
  }
```

This means that internal commands should respond in host errors.
So, the problem is probably because we don't have the step that moves through the errno encode/decode sequence due to using a UNIX domain socket?

Reply:
Oh, good lord. The application of that conversion is not universal in Ceph.

That particular change is something you made yourself, right? Although I see that `MCommandReply` already uses errorcode32_t which is not a struct I was familiar with. That does the encode/decode magic for you.

I think there needs to be more discussion in standup about this. We are not doing a good job being consistent with the rest of Ceph here. Part of the problem is that `MClientReply` standardized implicitly on Linux errnos. It does not use `errorcode32_t`. But maybe we can drop-in `errorcode32_t`.

Reply:
`MClientReply` already does this here for get

```
  int get_result() const {
    #ifdef _WIN32
    // libclient and libcephfs return CEPHFS_E* errors, which are basically
    // Linux errno codes. If we convert mds errors to host errno values, we
    // end up mixing error codes.
    //
    // For Windows, we'll preserve the original error value, which is expected
    // to be a linux (CEPHFS_E*) error. It may be worth doing the same for
    // other platforms.
    return head.result;
    #else
    return ceph_to_hostos_errno((__s32)(__u32)head.result);
    #endif
  }
```

but it doesn't do the opposite for the set:

```
  void set_result(int r) { head.result = r; }
```

I think that the latter should be 

```
  void set_result(int r) { head.result = hostos_to_ceph_errno(r); }
```

Reply:
I only found 4 invocations of the function in the mds code, and all four are consistent in that they pass the `CEPH_` errno to the set_result:

```
src/mds/Server.cc:
  446      dout(10) << __func__ << " invalid message (no uuid)" << dendl;
  447:     reply->set_result(-CEPHFS_EINVAL);
  448      mds->send_message_client(reply, session);

  454      dout(10) << __func__ << " unsupported flags" << dendl;
  455:     reply->set_result(-CEPHFS_EINVAL);
  456      mds->send_message_client(reply, session);

  464  	       << " != target auth_name " << target->info.auth_name << dendl;
  465:       reply->set_result(-CEPHFS_EPERM);
  466        mds->send_message_client(reply, session);

  545        auto reply = make_message<MClientReclaimReply>(0);
  546:       reply->set_result(-CEPHFS_EINVAL);
  547        mds->send_message_client(reply, session);
```

Reply:
OK scratch that, you are right, it's all over the place - what is returned by the ClientReply is supposed to be initialized by a CEPHFS_ errno. And I think that's the case almost everywhere.

However, if we do it all right, I'd still expect that the python test should be able to properly compare the rc to the host errno. The conversion should happen somewhere - in the client reply, it would be in the `get_result`

Reply:
So I think for now adding `CEPHFS_EINPROGRESS` is the right thing and consistent with the MDS today. Please make a tracker ticket for us to refactor/cleanup the error codes in the MDS.